Implement the IDL security check

yvt opened the pull request:
<https://heycam.github.io/webidl/#es-operations>
<https://html.spec.whatwg.org/multipage/#integration-with-idl>

This PR implements the IDL security check, which is an important (but not only) mechanism to restrict cross-origin DOM accesses.

This implementation follows WebKit's behavior and not the specification for the following reasons:

 - Neither Gecko nor WebKit implements the specification's behavior.
 - This would require a relatively elaborate mechanism *just* to have access to the target object's `CrossOriginProperties`.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

---
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

yvt committed:
test: test the IDL security check

<https://heycam.github.io/webidl/#es-operations>
<https://html.spec.whatwg.org/multipage/#integration-with-idl>

This security check is an important (but not only) mechanism to restrict
cross-origin DOM accesses.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

yvt committed:
refactor(script): add `CallPolicy` to bundle IDL operation customization parameters

This trait's associated constant `INFO` provides fields that control
various details of an IDL operation, such as whether a
`[LegacyLenientThis]` attribute is specified on the corresponding
interface member.

The `generic_(method|getter|setter)` functions are now generic over
`Policy: CallPolicy`, making the `generic_lenient_*` methods
unnecessary. Currently, they pass `Policy::INFO` to the inner
`generic_method` function as a runtime parameter, but we can change it
later if we find it beneficial to make generic the inner function.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

yvt committed:
test: update expectation

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/bindings/utils.rs, components/script/dom/bindings/proxyhandler.rs
 * @kichjang: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/bindings/utils.rs, components/script/dom/bindings/proxyhandler.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

yvt commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -1923,10 +1930,20 @@ def getter(attr):
-                    accessor = "generic_lenient_getter"
+                if attr.getExtendedAttribute("CrossOriginReadable"):
+                    assert not attr.hasLenientThis(), \
+                        "CrossOriginWritable && LenientThis: not supported"
```
Oops, I did [it](https://github.com/servo/servo/pull/28581) again!

yvt committed:
feat(script): implement the IDL security check with a WebKit-like behavior

<https://heycam.github.io/webidl/#es-operations>
<https://html.spec.whatwg.org/multipage/#integration-with-idl>

Not implementing the specification's behavior for the following reasons:

 - Neither Gecko nor WebKit implements this behavior.
 - This would require an elaborate mechanism to have access to the
   target object's `CrossOriginProperties`.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

bors-servo commented:
:umbrella: The latest upstream changes (presumably #29384) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
This is great, and I love the CallPolicy additions. I'm going to rebase these changes and get them merged.

jdm committed:
Lint fixes.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm committed:
Tidy.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

delan commented:
hi! i’m working on the monthly update. i did a bunch of testing to understand exactly what this patch fixes and what it doesn’t fix, but i’m still not 100% sure. it seems like before this patch, cross-site DOM accesses were already correctly rejected by servo, but same-site cross-origin DOM accesses were incorrectly allowed by servo. and after this patch, we fix one kind of incorrect access, specifically those that involve binding a Window or Location method in this document with a `this` from the other document. is that correct?

```html
<!--
    1. save this file as test.html
    2. serve the directory with both `python -m http.server 8000` and `8001`
    3. go to <http://localhost:8000/test.html>
-->
<!-- case 1: same-origin -->
<!-- <iframe id=i src="http://localhost:8000/"></iframe> -->
<!-- case 2: same-site cross-origin -->
<iframe id=i src="http://localhost:8001/"></iframe>
<!-- case 3: cross-site -->
<!-- <iframe id=i src="https://servo.org/"></iframe> -->
<script>
    setTimeout(()=>{
        try {
            // for case 2...
            // servo 0.1.0: returns `"#document"`
            // servo 0.2.0: returns `"#document"`
            // firefox 150: throws DOMException
            alert(i.contentWindow.document.nodeName);
        } catch (e) {
            console.log(e);
        }
        try {
            // for case 2...
            // servo 0.1.0: returns `undefined`
            // servo 0.2.0: returns `undefined`
            // firefox 150: throws DOMException
            alert(i.contentWindow.stop());
        } catch (e) {
            console.log(e);
        }
        try {
            // for case 2...
            // servo 0.1.0: returns `undefined`
            // servo 0.2.0: throws SecurityError
            // firefox 150: throws DOMException
            const d = Object.getOwnPropertyDescriptor(window,"stop");
            alert(d.value.bind(i.contentWindow)());
        } catch (e) {
            console.log(e);
        }
    },250)
</script>
```

jdm commented:
That is correct. This only affects the exotic WebIDL interfaces, Window and Location. #44649 is another big missing piece here.
